jupyterhub/scopes.py

```diff
@@ -13,60 +15,77 @@ class Scope(Enum):
+    parsed_scopes_b = parse_scopes(scopes_b)
 
-    common_bases = owner_parsed_scopes.keys() & token_parsed_scopes.keys()
+    common_bases = parsed_scopes_a.keys() & parsed_scopes_b.keys()
 
     common_filters = {}
-    warn = False
     for base in common_bases:
-        if owner_parsed_scopes[base] == Scope.ALL:
-            common_filters[base] = token_parsed_scopes[base]
-        elif token_parsed_scopes[base] == Scope.ALL:
-            common_filters[base] = owner_parsed_scopes[base]
+        filters_a = parsed_scopes_a[base]
+        filters_b = parsed_scopes_b[base]
+        if filters_a == Scope.ALL:
+            common_filters[base] = filters_b
+        elif filters_b == Scope.ALL:
+            common_filters[base] = filters_a
         else:
-            common_entities = (
-                owner_parsed_scopes[base].keys() & token_parsed_scopes[base].keys()
-            )
-            all_entities = (
-                owner_parsed_scopes[base].keys() | token_parsed_scopes[base].keys()
-            )
-            if 'user' in all_entities and ('group' or 'server' in all_entities):
-                warn = True
+            # warn *if* there are non-overlapping user= and group= filters
+            common_entities = filters_a.keys() & filters_b.keys()
+            all_entities = filters_a.keys() | filters_b.keys()
+            if 'group' in all_entities and 'user' in all_entities:
+                # this could resolve wrong if there's a user only on one side and a group only on the other
+                # check both directions: A has group X not in B group list AND B has user Y not in A user list
+                for a, b in [(filters_a, filters_b), (filters_b, filters_a)]:
+                    if (
+                        "group" in a
+                        and "user" in b
+                        and set(a["group"]).difference(b.get("group", []))
+                        and set(b["user"]).difference(a.get("user", []))
+                    ):
+                        warnings.warn(
+                            f"{base}[!user={b['user']}, !group={a['group']}] combinations of filters present,"
+                            " intersection between not considered. May result in lower than intended permissions.",
+                            UserWarning,
+                        )
 
             common_filters[base] = {
-                entity: set(owner_parsed_scopes[base][entity])
-                & set(token_parsed_scopes[base][entity])
+                entity: set(parsed_scopes_a[base][entity])
+                & set(parsed_scopes_b[base][entity])
                 for entity in common_entities
             }
 
-    if warn:
-        app_log.warning(
-            "[!user=, !group=] or [!user=, !server=] combinations of filters present, intersection between not considered. May result in lower than intended permissions."
-        )
+            if 'server' in all_entities and 'user' in all_entities:
```

Comment:
What if we have a `[read:users:servers!group=A, 'read:users:servers!server=abc/xyz]` where abc is member of group A? Technically that should be allowed, shouldn't it? Is it reasonable to raise a warning here as well for such situation?

Reply:
Yes, the group->server relationship is the same as group->user. Good catch!

Reply:
added tests and warning conditions for this